src/Test/Hspec/Expectations/Contrib.hs

```diff
@@ -24,3 +30,36 @@ isRight :: Either a b -> Bool
 isRight (Left  _) = False
 isRight (Right _) = True
 #endif
+
+-- |
+-- If you have a test case that has multiple assertions, you can use the
+-- 'annotate' function to provide a string message that will be attached to
+-- the 'Expectation'.
+--
+-- @
+-- describe "annotate" $ do
+--   it "adds the message" $ do
+--     annotate "obvious falsehood" $ do
+--       True `shouldBe` False
+--
+-- ========>
+--
+-- 1) annotate, adds the message
+--       obvious falsehood
+--       expected: False
+--        but got: True
+-- @
+--
+-- @since 0.8.3
+annotate :: String -> IO a -> IO a
```

Comment:
I generalized the type signature. At work we've got a few expectations of the form:

```haskell
assertJust :: Maybe a -> IO a
assertJust = \case
  Nothing -> expectationFailure "Expected Just, got Nothing" >> undefined
 Just a -> pure a
```

and the `Expectation ~ IO ()` is a problem